File: PatternCheck.java
//Time complexity = O(n)
//Space complexity = O(1)
/*Approach
*going through each of the char in pattern string and putting its correspondence in the hashmap
*if key is not present but the value is present then it will return false because one key can correspond to one
string in str
* if key is present we will match its correspondence in the hashmap as well as in that position(i) if they dont match
we will return false
* if no false case becomes true we will return true
*/
import java.util.HashMap;
import java.util.Map;
public class PatternCheck {
    public static boolean wordPattern(String pattern, String str) {
        Map<Character, String> map = new HashMap<>();
        String[] strArray = str.split(" ");
        char[] patternArray = pattern.toCharArray();
        if(strArray.length != patternArray.length) return false;
        for(int i = 0; i < patternArray.length;i++){
            if(!map.containsKey(patternArray[i])){
                if(map.containsValue(strArray[i])) return false;
                map.put(patternArray[i],strArray[i]);
            }
            else{
                if(!map.get(patternArray[i]).equals(strArray[i])){
                    return false;
                }
            }
        }
        
     return true;
    }
    public static void main(String args[]){
        System.out.println(wordPattern("abba","fish ant ant fish"));
        System.out.println(wordPattern("abba","fish fish fish fish"));
    }

}

File: patternCheck.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        letter_2_word = {}
        word_2_letter = {}
        if len(pattern) != len(s.split()):
            return False
        for c, word in zip(pattern, s.split()):
            letter_2_word[c] = letter_2_word.get(c, word)
            word_2_letter[word] = word_2_letter.get(word, c)
            if letter_2_word[c] != word or word_2_letter[word] != c:
                return False
        return True
        